packages/commerce-server/src/determine-purchase-type.ts

```diff
@@ -0,0 +1,121 @@
+import {isEmpty} from 'lodash'
+import {getSdk, Context, defaultContext} from '@skillrecordings/database'
+import {Context as StripeContext} from '@skillrecordings/stripe-sdk'
+import {stripeData} from './record-new-purchase'
+import {
+  EXISTING_BULK_COUPON,
+  NEW_BULK_COUPON,
+  NEW_INDIVIDUAL_PURCHASE,
+  INDIVIDUAL_TO_BULK_UPGRADE,
+} from '@skillrecordings/types'
+import {z} from 'zod'
+
+export const purchaseTypeSchema = z.union([
+  z.literal(EXISTING_BULK_COUPON),
+  z.literal(NEW_BULK_COUPON),
+  z.literal(NEW_INDIVIDUAL_PURCHASE),
+  z.literal(INDIVIDUAL_TO_BULK_UPGRADE),
+])
+export type PurchaseType = z.infer<typeof purchaseTypeSchema>
+
+/**
+ * Check that two values match without matching on two undefined values
+ *
+ * This is helpful when you find yourself doing a comparison like
+ * `obj?.a === obj?.b`. If both values are undefined, then it resolves to
+ * true. If you don't want that, then you have to guard the comparison,
+ * `obj?.a && obj?.b && obj?.a === obj?.b`. This function takes care of that.
+ */
+const match = (
+  value1: string | number | undefined,
+  value2: string | number | undefined,
+) => {
+  return Boolean(value1 && value2 && value1 === value2)
+}
+
+type DeterminePurchaseTypeOptions = {
+  checkoutSessionId: string
+  prismaCtx?: Context
+  stripeCtx?: StripeContext
+}
+
+export async function determinePurchaseType(
```

Comment:
This is the headline function/module -- there were a couple places that were running bespoke queries and lines of logic to determine the type of purchase that had been made given a checkout session ID.

This function is now the source of truth for making that determination.

It also uses the `packages/database` `getSdk` for all prisma calls and `packages/stripe-sdk` for the Stripe call. This makes it testable with mocks (see the accompanying test file).